File: 031/cpp/main.cpp
class Solution {
public:
    void swap(vector<int>& nums, int i, int j) {
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }

    bool swap(vector<int>& nums, int i) {
        if (i == nums.size() - 1) {
            return false;
        }
        if (swap(nums, i + 1)) {
            return true;
        }
        bool r = false;
        for (int j = nums.size() - 1; j > i; j--) {
            if (nums[j] > nums[i]) {
                swap(nums, i, j);
                r = true;
                break;
            }
        }
        if (!r) {
            return false;
        }
        std::reverse(nums.begin() + i + 1, nums.end());
        return true;
    }

    void nextPermutation(vector<int>& nums) {
        if (!swap(nums, 0)) {
            std::reverse(nums.begin(), nums.end());
        }
    }
};

File: 031/rust/main.rs
impl Solution {
    pub fn reverse(nums: &mut Vec<i32>, i: usize) {
        let n = nums.len();
        let m = nums.len() - i;
        for j in 0..m / 2 {
            nums.swap(i + j, n - 1 - j);
        }
    }

    pub fn swap(nums: &mut Vec<i32>, i: usize) -> bool {
        if i == nums.len() - 1 {
            return false;
        }
        if Solution::swap(nums, i + 1) {
            return true;
        }
        let mut r = false;
        for j in (i + 1..nums.len()).rev() {
            if nums[j] > nums[i] {
                nums.swap(i, j);
                r = true;       
                break;
            }
        }
        if !r {
            return false;
        }
        Solution::reverse(nums, i + 1);
        return true;
    }

    pub fn next_permutation(nums: &mut Vec<i32>) {
        let r = Solution::swap(nums, 0);
        if !r {
            nums.reverse();
        }
    }
}
